libr/bin/p/bin_nso.c

```diff
@@ -0,0 +1,269 @@
+/* radare2 - LGPL - Copyright 2017 - rkx1209 */
+
+#include <r_types.h>
+#include <r_util.h>
+#include <r_lib.h>
+#include <r_bin.h>
+#include <r_io.h>
+#include <r_cons.h>
+#include "../../../shlr/lz4/lz4.h"
+
+#define NSO_OFF(x) r_offsetof (NSOHeader, x)
+
+// starting at 0
+typedef struct {
+	ut32 magic;	// NSO0
+	ut32 pad0;	// 4
+	ut32 pad1;	// 8
+	ut32 pad2;	// 12
+	ut32 text_memoffset;	// 16
+	ut32 text_loc;	// 20
+	ut32 text_size;	// 24
+	ut32 pad3;	// 28
+	ut32 ro_memoffset;	// 32
+	ut32 ro_loc;	// 36
+	ut32 ro_size;	// 40
+	ut32 pad4;	// 44
+	ut32 data_memoffset;	// 48
+	ut32 data_loc;	// 52
+	ut32 data_size;	// 56
+	ut32 bss_size;	// 60
+} NSOHeader;
+
+static ut32 readLE32(RBuffer *buf, int off) {
+	int left = 0;
+	const ut8 *data = r_buf_get_at (buf, off, &left);
+	return left > 3? r_read_le32 (data): 0;
+}
+
+static uint32_t decompress(const ut8 *cbuf, ut8 *obuf, uint32_t csize, uint32_t usize) {
+	return LZ4_decompress_safe (cbuf, obuf, csize, usize);
+}
+
+static ut64 baddr(RBinFile *bf) {
+	return 0;	// XXX
+}
+
+static const char *fileType(const ut8 *buf) {
+	if (!memcmp (buf, "NSO0", 4)) {
+		return "nso0";
+	}
+	return NULL;
+}
+
+static bool check_bytes(const ut8 *buf, ut64 length) {
+	if (buf && length >= 0x20) {
+		return fileType (buf + NSO_OFF (magic)) != NULL;
+	}
+	return false;
+}
+
+static void *load_bytes(RBinFile *bf, const ut8 *buf, ut64 sz, ut64 loadaddr, Sdb *sdb) {
+	RBin *rbin = bf->rbin;
+	ut32 toff = readLE32 (bf->buf, NSO_OFF (text_memoffset));
+	ut32 tsize = readLE32 (bf->buf, NSO_OFF (text_size));
+	ut32 rooff = readLE32 (bf->buf, NSO_OFF (ro_memoffset));
+	ut32 rosize = readLE32 (bf->buf, NSO_OFF (ro_size));
+	ut32 doff = readLE32 (bf->buf, NSO_OFF (data_memoffset));
+	ut32 dsize = readLE32 (bf->buf, NSO_OFF (data_size));
+	ut64 total_size = tsize + rosize + dsize;
+	ut8 *newbuf = malloc (total_size);
```

Comment:
check if malloc succeeds or not, may we use calloc here instead? is decompress filling out this newbuf completely?